backend/x86_peephole/x86_peephole_rules.ml

```diff
@@ -0,0 +1,437 @@
+[@@@ocaml.warning "+a-29-40-41-42-4"]
+
+open! Int_replace_polymorphic_compare
+open X86_ast
+module DLL = Oxcaml_utils.Doubly_linked_list
+module U = X86_peephole_utils
+
+type peephole_stats =
+  { mutable remove_mov_x_x : int;
+    mutable remove_useless_mov : int;
+    mutable remove_mov_chain : int;
+    mutable remove_mov_to_dead_register : int;
+    mutable rewrite_mov_sequence_to_xchg : int;
+    mutable rewrite_mov_add_to_lea : int;
+    mutable rewrite_mov_add_reg_to_lea : int;
+    mutable remove_redundant_cmp : int;
+    mutable combine_add_rsp : int
+  }
+
+let create_peephole_stats () =
+  { remove_mov_x_x = 0;
+    remove_useless_mov = 0;
+    remove_mov_chain = 0;
+    remove_mov_to_dead_register = 0;
+    rewrite_mov_sequence_to_xchg = 0;
+    rewrite_mov_add_to_lea = 0;
+    rewrite_mov_add_reg_to_lea = 0;
+    remove_redundant_cmp = 0;
+    combine_add_rsp = 0
+  }
+
+let peephole_stats_to_counters stats =
+  Profile.Counters.create ()
+  |> Profile.Counters.set "x86_peephole.remove_mov_x_x" stats.remove_mov_x_x
+  |> Profile.Counters.set "x86_peephole.remove_useless_mov"
+       stats.remove_useless_mov
+  |> Profile.Counters.set "x86_peephole.remove_mov_chain" stats.remove_mov_chain
+  |> Profile.Counters.set "x86_peephole.remove_mov_to_dead_register"
+       stats.remove_mov_to_dead_register
+  |> Profile.Counters.set "x86_peephole.rewrite_mov_sequence_to_xchg"
+       stats.rewrite_mov_sequence_to_xchg
+  |> Profile.Counters.set "x86_peephole.rewrite_mov_add_to_lea"
+       stats.rewrite_mov_add_to_lea
+  |> Profile.Counters.set "x86_peephole.rewrite_mov_add_reg_to_lea"
+       stats.rewrite_mov_add_reg_to_lea
+  |> Profile.Counters.set "x86_peephole.remove_redundant_cmp"
+       stats.remove_redundant_cmp
+  |> Profile.Counters.set "x86_peephole.combine_add_rsp" stats.combine_add_rsp
+
+(* CR xclerc: check where the rules should set the "restart" cell to be sure we
+   always apply all rules. *)
+
+(* Rewrite rule: remove MOV x, x (moving a value to itself) Note: We can only
+   safely remove self-moves for registers that don't have zero-extension side
+   effects. On x86-64: - 32-bit moves (Reg32) zero the upper 32 bits - SIMD
+   moves (Regf) may zero upper bits depending on instruction encoding So we only
```

Comment:
I think all `Regf` moves now go through the SIMD instruction path, which distinguishes between zero-extending and preserving moves (e.g. `movss` vs `vmovss`). I don't think this PR needs to handle floats, but it could use a cr-soon